### PR TITLE
Refactor external dispatch actions from being called inside useSelect

### LINKED ISSUE #5866: Cart block not working on WordPress.com site
## Describe the bug

If we create a new page and add a cart block on a WordPress.com site, the screen gets freeze, and the browser gets crashes after a few minutes.

## To reproduce

1. Go to My Site → Pages → Add New. 
2. Add Cart block to the page.  
3. Notice the screen gets freeze. 

## Expected behavior

Cart block should be added to the page 

### WordPress:

-   Core version: 5.9
-   Gutenberg version (if installed): 12.5.4 
-   WooCommerce version: 6.2
-   WooCommerce Blocks version: 6.9.0

### LINKED ISSUE #57449: WooCommerce Blocks: Adding the Cart block crashes the editor
### Update from @JoshuaGoode 

Please see internal notes p9F6qB-7LA-p2#comment-43978

Related
304-gh-Automattic/wpcomsh
#34655

### Quick summary

When adding the WooCommerce Cart block I get this error:

```
TypeError: Cannot read properties of null (reading 'removeChild')
    at h.componentWillUnmount (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/amp/assets/js/amp-block-editor.js?ver=2ee98517d90b7670631d2afc67df7285:31:1273)
    at Rh (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:148:43)
    at Sh (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:151:394)
    at Qj (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:174:181)
    at unstable_runWithPriority (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react.min.755b0ae9.js:24:26)
    at Za (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:73:8)
    at eb (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:170:163)
    at gf (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:162:317)
    at https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react-dom.min.edd8b7a6.js:73:230
    at unstable_runWithPriority (https://woocrashcourse101463901178.wpcomstaging.com/wp-content/plugins/gutenberg/vendor/react.min.755b0ae9.js:24:26)
```

### Steps to reproduce

1. On an AT Site install WooCommerce blocks: https://woocommerce.com/products/woocommerce-gutenberg-products-block/
2. Create a new page and add the WooCommerce Cart block
3. Notice the error

I confirmed this doesn't happen on a WordPress.org site (Jurassic Ninja)

### What you expected to happen

For the editor to display the Cart block

### What actually happened

Got this message:

>The editor has encountered an unexpected error.

https://user-images.githubusercontent.com/3696121/139307197-8a8c846b-6e78-4004-ae71-a7c815ebaf47.mov

Console error:

Screen Shot 2021-10-28 at 12 41 13

### Context

I encountered this while doing the WooCommerce Crash Course

### Operating System

macOS

### Browser

Google Chrome/Chromium

### Simple, Atomic or both?

Atomic

### Theme-specific issue?

_No response_

### Other notes

_No response_

### Reproducibility

Consistent

### Severity

Some (< 50%)

(WooCommerce Blocks are an experimental feature )

### Available workarounds?

_No response_

### Workaround details

_No response_

### PR DESCRIPTION
**Pull request by senadir**
This PR resolves the long-standing issue in wordpress.com in which Cart and Checkout blocks would cause an infinite loop and, recently, crash some blocks.

The source of the issue was that the wpcom [code that did the monkey patching](https://github.com/Automattic/wp-calypso/blob/trunk/apps/wpcom-block-editor/src/wpcom/features/tracking.js#L767-L800) couldn't handle actions being called from inside a useSelect or useDispatch function, causing an infinite loop in Gutenberg.

I couldn't really pinpoint the exact line that handle that, maybe someone more experienced in that codebase can look cc @Addison-Stavlo.

For this PR, I just moved the logic from outside the useSelect call, it made sense anyway to do that outside.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/5866
Fixes https://github.com/Automattic/wp-calypso/issues/57449

### Note
Because we didn't make the wpcom code more resilient yet, this issue might surface back again, if it does, the fix would be to simply create your functions outside the useSelect function, not inside it.

### Testing

To test this PR, you will need to have a zip to upload to an atomic website `*.wpcomstaging.com`. Use this zip 
[woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/9144395/woocommerce-gutenberg-products-block.zip) that reflects this PR.

1. Create a new atomic website (wordpress.com -> new website -> pro plan -> install WooCommerce).
2. Upload and enable this plugin.
3. Open the editor, open the console as well.
4. Insert Cart block, you shouldn't see any console errors, all block should load fine without errors.
5. Insert Checkout block, you shouldn't see any console errors, all blocks should load fine without errors.
6. Try adding a product, going to cart and checkout, they should load fine.

Outside wpcom, make sure this PR doesn't regress anything:
- Create a coupon.
- Go to cart, try to use the coupon, you should see a snackbar below, no errors in console.
- Go to checkout, remove the coupon, insert it again, no error should surface.

* [ ] Do not include in the Testing Notes: Only include the second part of testing.

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Fix Cart an d Checkout blocks compatiblity issue with wordpress.com in which blocks wouldn't load in the editor.

### CODE REVIEW

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (senadir) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:43]: This is a broken TS, wasn't sure what's the best fix @opr either add `receiveApplyingCoupon` to `StoreCartCoupon` or have another sort of union. Not even sure what `StoreCartCoupon` should be and what should it include.

**Comment by Addison-Stavlo** [comment]
👋  Thanks for unblocking this here.

> Because we didn't make the wpcom code more resilient yet, this issue might surface back again,

Id be happy to see what we can do to resolve this. Im having a hard time understanding the cause here though.

>  The source of the issue was that the wpcom [code that did the monkey patching](https://github.com/Automattic/wp-calypso/blob/trunk/apps/wpcom-block-editor/src/wpcom/features/tracking.js#L767-L800) couldn't handle actions being called from inside a useSelect or useDispatch function, causing an infinite loop in Gutenberg.

Just looking at the changes here, it looks like we weren't previously calling these actions within the `useSelect` here though? Just defining other functions that use them when called elsewhere?

On another note, ive never actually thought about calling `dispatch` from within `useSelect`. Generally I thought standard practice would be to separate the two (as is done with [this example of using useSelect in combination with useDispatch](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#usedispatch) from the block editor reference handbook)

```
const { isApplyingCoupon, isRemovingCoupon } = useSelect( (select) => ...  );
const { applyCoupon, removeCoupon, receiveApplyingCoupon } = useDispatch( storeName );
// then using the above to define further functions/logic
```
Is pulling dispatch actions out of `useSelect` a common practice?

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (opr) [reply assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:43]: I don't think we need to type this, it should be inferred, and it also seems weird to still pick dispatch actions inside `useSelect` I propose we drop the explicit typing, remove the `actions` from the `useSelect` and get them from `useDispatch` instead. This is consistent with how it's being done elsewhere in the codebase and the data store refactor work.

I tested this on wpcom and it works still.

See this example patch and let me know what you think.

```patch
From c7787796b798e7914f290c03f484b9fd19f99a7f Mon Sep 17 00:00:00 2001
From: Thomas Roberts <thomas.roberts@automattic.com>
Date: Fri, 22 Jul 2022 09:35:31 +0100
Subject: [PATCH] Add new dispatch outside of select

---
 .../hooks/cart/use-store-cart-coupons.ts      | 23 ++++---------------
 1 file changed, 5 insertions(+), 18 deletions(-)

diff --git a/assets/js/base/context/hooks/cart/use-store-cart-coupons.ts b/assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
index aa5cd0bf8..2115fd330 100644
--- a/assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
+++ b/assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
@@ -29,34 +29,21 @@ export const useStoreCartCoupons = ( context = '' ): StoreCartCoupon => {
 	const { createNotice } = useDispatch( 'core/notices' );
 	const { setValidationErrors } = useValidationContext();
 
-	const {
-		applyCoupon,
-		removeCoupon,
-		isApplyingCoupon,
-		isRemovingCoupon,
-	}: Pick<
-		StoreCartCoupon,
-		| 'applyCoupon'
-		| 'removeCoupon'
-		| 'isApplyingCoupon'
-		| 'isRemovingCoupon'
-		| 'receiveApplyingCoupon'
-	> = useSelect(
-		( select, { dispatch } ) => {
+	const { isApplyingCoupon, isRemovingCoupon } = useSelect(
+		( select ) => {
 			const store = select( storeKey );
-			const actions = dispatch( storeKey );
 
 			return {
-				applyCoupon: actions.applyCoupon,
-				removeCoupon: actions.removeCoupon,
 				isApplyingCoupon: store.isApplyingCoupon(),
 				isRemovingCoupon: store.isRemovingCoupon(),
-				receiveApplyingCoupon: actions.receiveApplyingCoupon,
 			};
 		},
 		[ createErrorNotice, createNotice ]
 	);
 
+	const { applyCoupon, removeCoupon, receiveApplyingCoupon } =
+		useDispatch( storeKey );
+
 	const applyCouponWithNotices = ( couponCode: string ) => {
 		applyCoupon( couponCode )
 			.then( ( result ) => {
-- 
2.32.1 (Apple Git-133)
```

**Review by opr — DISMISSED** [review]

**Review by nielslange — APPROVED** [review]
While @opr inspected the code, I did a smoke test and kept an eye on the console. The PR works as expected, given the areas I inspected. Once we have @opr's approval, let's ⛴ this fix.

**Comment by senadir** [comment]
Hey @Addison-Stavlo 

> Id be happy to see what we can do to resolve this. Im having a hard time understanding the cause here though.

I'd lie if I said I fully understand what's happening, I think this might be a scope issue, but I would expect a visible error instead of useSelect looping at the `onStoreChange` call.

> Just looking at the changes here, it looks like we weren't previously calling these actions within the useSelect here though? Just defining other functions that use them when called elsewhere?

Yes, I might have worded this badly, but we're just defining a function that will call them outside the useSelect, but the issue happen if you define the action variable inside the useSelect and construct it there.

> On another note, ive never actually thought about calling dispatch from within useSelect. Generally I thought standard practice would be to separate the two (as is done with [this example of using useSelect in combination with useDispatch](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#usedispatch) from the block editor reference handbook)

It is a shortcode used to save code, y[ou can find usage of that in Gutenberg](https://github.com/WordPress/gutenberg/search?p=1&q=%22select%2C+dispatch%22), for our use case, we don't gain much so it's fine to remove it.

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (senadir) [reply assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:43]: Thank you Thomas! I applied the changes.

**Comment by Addison-Stavlo** [comment]
> It is a shortcode used to save code, y[ou can find usage of that in Gutenberg](https://github.com/WordPress/gutenberg/search?p=1&q=%22select%2C+dispatch%22), for our use case, we don't gain much so it's fine to remove it.

Hrm, I don't see any of those results reflecting what was being done here in terms of destructuring dispatch out of the store in useSelect - `useSelect( ( select, { dispatch } ) =>`

> I'd lie if I said I fully understand what's happening, I think this might be a scope issue, but I would expect a visible error instead of useSelect looping at the onStoreChange call.

Yeah, Im glad you have something working but am definitely confused here as well. I haven't seen any other blocks crashing the editor like this, and originally found `insertBlock(s)` and `replaceInnerBlocks` to be the problematic actions here from the tracking side (adding any handling fn to those actions, even noop, caused the crash combined with the cart block - removing them entirely resolved it) - but this fix seems like it has nothing to do with those actions. 🤷‍♀️ 🤔 Wondering if these things somehow effected each other.

**Comment by senadir** [comment]
> Yeah, Im glad you have something working but am definitely confused here as well. I haven't seen any other blocks crashing the editor like this, and originally found insertBlock(s) and replaceInnerBlocks to be the problematic actions here from the tracking side (adding any handling fn to those actions, even noop, caused the crash combined with the cart block - removing them entirely resolved it) - but this fix seems like it has nothing to do with those actions. 🤷‍♀️ 🤔 Wondering if these things somehow effected each other.

Something changed in Gutenberg around that because we didn't change anything worthwhile in our code nor did you, however that change did surface other the same error but in an easier to debug state.

> Hrm, I don't see any of those results reflecting what was being done here in terms of destructuring dispatch out of the store in useSelect - useSelect( ( select, { dispatch } ) =>

It's true, it seems the current usage is only for creating actions, not inside the useSelect, it must have been refactored away because I do remember this pattern (and it's from GB that I picked it up).

I asked in slack about that behavior but got no responses, I will keep an eye for it so it doens't surface again.

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (nerrad) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:51]: Could just move this out of the `useSelect` callback too? (same with `receiveApplyingCoupon`)

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (nerrad) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:57]: These dependencies shouldn't be needed now?

File: assets/js/base/context/hooks/cart/use-store-cart-coupons.ts
Comment (nerrad) [inline assets/js/base/context/hooks/cart/use-store-cart-coupons.ts:60]: I don't know how often `createNotice` or `setValidationErrors` or `createErrorNotice` change (can't remember if they themselves are dispatch function) but one optimization here might be to memoize the new composite dispatchers via `useCallback`. I _think_ that might have been why they were in the `useSelect` callback originally so that we memoized them based on the dependencies.

**Pull request merged**
